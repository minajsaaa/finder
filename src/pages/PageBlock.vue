<template lang="pug">
  div
    app-header
    div(v-if="!block.loaded && block.loading")
      app-loading
    div(class="block-container" v-else-if="block.loaded && !block.loading && !isEmpty(blockData.header)")
      h2 Block
        span {{ `#${blockData.header.height}` }}

      div(class="table")
        tm-list-item(dt="Block height", class="height")
          template(slot="dd")
            span {{ blockData.header.height }}
            router-link(v-if="blockData.header.height > 1" :to="{ name: 'block', params: { block: parseInt(blockData.header.height) - 1 }}"): i.material-icons chevron_left
            router-link(:to="{ name: 'block', params: { block: parseInt(blockData.header.height) + 1 }}"): i.material-icons chevron_right
        tm-list-item(dt="Timestamp" :dd="`${format(blockData.header.time)} (Local Time)`")
        tm-list-item(dt="Transactions")
          template(slot="dd")
            router-link(v-if="parseInt(blockData.header.num_txs) > 0" class="transactions" :to="{ name: 'txs', params: { block: parseInt(blockData.header.height) }}")
              span {{ blockData.header.num_txs }} Transactions
                i.material-icons keyboard_arrow_right
            a(v-else class="transactions none")
              span {{ blockData.header.num_txs }} Transactions

        tm-list-item(dt="Proposer")
          template(slot="dd")
            span {{ blockData.header.proposer_address }}
        //- tm-list-item(dt="Block reward" :dd="`0.8888 luna`")
        //- tm-list-item(dt="Size" :dd="`8,888 bytes`")
        tm-list-item(dt="Block hash" :dd="blockMeta.block_id.hash")
        tm-list-item(dt="Parent hash")
          template(slot="dd")
            router-link(:to="{ name: 'block', params: { block: parseInt(blockData.header.height) - 1 }}") {{ blockData.last_commit.block_id.hash }}

    template(v-else-if="block.error && !block.loading && block.loaded")
      app-not-found
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import { isEmpty } from "lodash";
import { format } from "../scripts/utility";
import TmListItem from "../components/TmListItem";

import AppHeader from "../components/AppHeader";
import AppNotFound from "../components/AppNotFound";
import AppLoading from "../components/AppLoading";

export default {
  beforeCreate: function() {
    document.body.className = "page";
  },
  name: "page-block",
  components: {
    TmListItem,
    AppHeader,
    AppNotFound,
    AppLoading
  },
  computed: {
    ...mapGetters(["block"]),
    blockData() {
      return (
        (this.block.blocks[this.$route.params.block] &&
          this.block.blocks[this.$route.params.block].block) || {
          last_commit: {
            block_id: {}
          },
          header: {}
        }
      );
    },
    blockMeta() {
      return (
        (this.block.blocks[this.$route.params.block] &&
          this.block.blocks[this.$route.params.block].block_meta) || {
          block_id: {
            hash: ``
          }
        }
      );
    }
  },
  async created() {
    await this.setBlockLoadedFalse();
    await this.fetchBlock(this.$route.params);
  },
  methods: {
    ...mapActions(["fetchBlock", "setBlockLoadedFalse"]),
    format,
    isEmpty
  },
  watch: {
    // eslint-disable-next-line
    $route(to, from) {
      this.setBlockLoadedFalse();
      this.fetchBlock(this.$route.params);
    }
  }
};
</script>

<style lang="stylus">
@require '../styles/variables'
.block-container
  width 100%
  max-width 1440px
  margin 0 auto
  padding 40px
  color #2043b5 !important

.block-container h2
  font-size 24px
  font-weight 500
  margin-bottom 20px

.block-container h2 span
  font-weight 300
  font-size 23px
  padding-left 6px

.block-container .table
  border solid 1px #d8ddf0
  border-radius 5px
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);

.block-container .table .tm-li-dl
  display table
  table-layout fixed
  width 100%
  min-height 60px

.block-container .table .tm-li-dl .tm-li-dt
  display table-cell
  vertical-align middle

.block-container .table .tm-li-dl .tm-li-dd
  display table-cell
  padding 15px 30px
  vertical-align middle
  white-space normal
  text-overflow unset
  word-break break-all
  line-height 1.5

.block-container a.transactions
  display inline-block
  position relative
  cursor pointer
  font-size 15px
  box-shadow 0 1px 2px 0 rgba(0, 0, 0, 0.1)
  line-height 30px
  height 30px
  border-radius 15px
  background-color #5493f7
  padding-left 20px
  padding-right: 46px
  color #ffffff
  text-decoration none
  transition background-color .1s
  white-space nowrap


.block-container a.transactions:hover
  text-decoration none
  background-color #2845AE

.block-container a.none
  background-color rgba(84, 147, 247,.5);
  padding 0 20px
  cursor default


.block-container .transactions i
  margin-left 15px
  font-size 20px
  position: absolute
  top: 6px

.block-container .height span
  margin-right 10px
  display inline-block
  vertical-align middle

.block-container .height a
  position relative
  width 22px
  height 22px
  display inline-block
  color #2845AE
  vertical-align middle
  margin-top -2px
  margin-right 6px;
  border-radius 50%
  background-color #edf4fe
  text-align center
  transition .1s

.block-container .height a:hover
  color #fff
  background #2845AE

.block-container .height a i
  line-height 22px

@media screen and (max-width: 767px)
  .block-container
    padding 50px 15px

  .block-container h2
    padding 40px 0 30px
    border-bottom 1px solid #d8ddf0
    margin 0

  .block-container .table
    box-shadow none
    border 0

  .block-container .table .tm-li-dl
    display block
    padding 0
    height auto
    line-height 1.5

  .block-container .table .tm-li-dl .tm-li-dt
    display block
    background transparent
    padding 20px 0 10px
    word-wrap break-word
    white-space unset
    font-size 15px

  .block-container .table .tm-li-dl .tm-li-dd, .tm-li-dd.tm-li-dd-flush
    display block
    padding 0 0 20px
    font-size 15px
    word-wrap break-word
    white-space unset

  .block-container a.transactions
    font-size 14px

  .block-container .tm-li
    border-bottom 1px solid #d8ddf0
</style>
