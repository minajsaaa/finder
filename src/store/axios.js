import axios from "axios";

const instance = axios.create({
  timeout: 15000
});
export default instance;
