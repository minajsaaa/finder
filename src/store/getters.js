export const block = state => state.block;
export const config = state => state.config;
export const tx = state => state.tx;
export const account = state => state.account;
