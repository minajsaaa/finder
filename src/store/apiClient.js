import axios from "axios";
import { getNetwork } from "../networks";

const limit = 1000;

function lcdUrl(network) {
  return getNetwork(network).lcd;
}

const instance = axios.create({
  timeout: 15000
});

const getBlock = (network, blockHeight, requester = instance) => {
  let url = `${lcdUrl(network)}/blocks/${blockHeight}`;
  return requester.get(url);
};

const getTx = (network, hash, requester = instance) => {
  let url = `${lcdUrl(network)}/txs/${hash.toUpperCase()}`;
  return requester.get(url);
};

const getAccount = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/auth/accounts/${address}`;
  return requester.get(url);
};

const getDelegations = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/staking/delegators/${address}/delegations`;
  return requester.get(url);
};

const getRewards = (
  network,
  delegatorAddress,
  validatorAddress,
  requester = instance
) => {
  let url = `${lcdUrl(
    network
  )}/distribution/delegators/${delegatorAddress}/rewards/${validatorAddress}`;
  return requester.get(url);
};

const getUndelegations = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/staking/delegators/${address}/unbonding_delegations`;
  return requester.get(url);
};

const getSenderTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/txs?sender=${address}&limit=${limit}`;
  return requester.get(url);
};
const getRecipientTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/txs?recipient=${address}&limit=${limit}`;
  return requester.get(url);
};
const getSwapTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/txs?action=swap&trader=${address}&limit=${limit}`;
  return requester.get(url);
};
const getProposerTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/txs?action=submit_proposal&proposer=${address}&limit=${limit}`;
  return requester.get(url);
};
const getDepositorTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/txs?action=deposit&depositor=${address}&limit=${limit}`;
  return requester.get(url);
};
const getDelegatorsTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/staking/delegators/${address}/txs?limit=${limit}`;
  return requester.get(url);
};

export default {
  getBlock,
  getTx,
  getAccount,
  getDelegations,
  getRewards,
  getUndelegations,
  getSenderTxs,
  getRecipientTxs,
  getSwapTxs,
  getProposerTxs,
  getDepositorTxs,
  getDelegatorsTxs
};
