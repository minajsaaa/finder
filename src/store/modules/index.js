import apiClient from "../apiClient";

const files = require.context(".", false, /\.js$/);
let modules = {};

files.keys().forEach(key => {
  if (key === "./index.js") return;
  modules[key.replace(/(\.\/|\.js)/g, "")] = files(key).default(apiClient);
});

export default modules;
