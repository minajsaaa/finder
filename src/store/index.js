import Vue from "vue";
import Vuex from "vuex";
import VueAnalytics from "vue-analytics";
// import * as actions from "./actions";
import * as getters from "./getters";
import modules from "./modules";
import * as Sentry from "@sentry/browser";
import * as Integrations from "@sentry/integrations";

Vue.use(VueAnalytics, {
  id: "UA-122601594-2"
});

Sentry.init({
  dsn: "https://c2041cd0a20e4c7cb6430f1472df397e@sentry.io/1463310",
  integrations: [
    new Integrations.Vue({
      Vue,
      attachProps: true
    })
  ]
});

Vue.use(Vuex);

export default new Vuex.Store({
  getters,
  modules,
  strict: false
});
