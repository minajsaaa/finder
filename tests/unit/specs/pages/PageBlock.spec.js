import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import * as getters from "@/store/getters";
import PageBlock from "@/pages/PageBlock";
import mockState from "../store/mockState";

describe(`PageBlock`, () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);
  let wrapper, $store, $route, mockStore, localThis;
  const defaultBlock = 42
  const errorBlock = 43
  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: {
        block: {}
      }
    };

    $route = {};

    wrapper = shallowMount(PageBlock, {
      localVue,
      mocks: {
        $store,
        $route
      }
    });
  });

  describe(`Snapshot`, () => {
    beforeEach(() => {
      mockStore = new Vuex.Store({
        state: mockState,
        getters,
        commit: jest.fn(),
        dispatch: jest.fn(),
        actions: {
          setBlockLoadedFalse: jest.fn(),
          fetchBlock: jest.fn()
        }
      });
    });

    it(`Snapshot: defaultBlock`, () => {
      wrapper = shallowMount(PageBlock, {
        localVue,
        store: mockStore,
        mocks: {
          $route: {
            params: {
              block: defaultBlock
            }
          }
        },
        stubs: [`router-link`]
      });
      expect(wrapper.vm.$el).toMatchSnapshot();
    });

    it(`Snapshot: load Error`, () => {
      wrapper = shallowMount(PageBlock, {
        localVue,
        mocks: {
          $store: {
            commit: jest.fn(),
            dispatch: jest.fn(),
            getters: {
              block: {
                blocks: {},
                loading: false,
                loaded: true,
                error: new Error(`err`)
              }
            }
          },
          $route: {
            params: {
              block: `99999999999999999`
            }
          },
          setBlockLoadedFalse: jest.fn(),
          fetchBlock: jest.fn()
        },
        stubs: [`router-link`]
      });
      expect(wrapper.vm.$el).toMatchSnapshot();
    });
  });

  describe(`Using localThis`, () => {
    beforeEach(() => {
      localThis = {
        $route: {
          params: {
            block: defaultBlock
          }
        },
        block: mockState.block,
        setBlockLoadedFalse: jest.fn(),
        fetchBlock: jest.fn()
      };
    });
    it(`Computed: blockData with defaultBlock`, () => {
      const blockHeight = localThis.$route.params.block;
      expect(PageBlock.computed.blockData.call(localThis)).toBe(
        mockState.block.blocks[blockHeight].block
      );
    });

    it(`Computed: no blockData with errorBlock`, () => {
      localThis.$route.params.block = errorBlock;

      expect(PageBlock.computed.blockData.call(localThis)).toEqual({
        last_commit: {
          block_id: {}
        },
        header: {}
      });
    });

    it(`Computed: blockMeta with defaultBlock`, () => {
      const blockHeight = localThis.$route.params.block;
      expect(PageBlock.computed.blockMeta.call(localThis)).toBe(
        mockState.block.blocks[blockHeight].block_meta
      );
    });

    it(`Computed: no blockMeta with errorBlock`, () => {
      localThis.$route.params.block = errorBlock;

      expect(PageBlock.computed.blockMeta.call(localThis)).toEqual({
        block_id: {
          hash: ``
        }
      });
    });

    it(`Watch: $route`, () => {
      PageBlock.watch[`$route`].call(localThis);

      expect(localThis.setBlockLoadedFalse).toHaveBeenCalled();
      expect(localThis.fetchBlock).toHaveBeenCalled();
    });
  });
});
