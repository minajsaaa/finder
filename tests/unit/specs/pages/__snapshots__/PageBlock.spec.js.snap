// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBlock Snapshot Snapshot: defaultBlock 1`] = `
<div>
  <app-header-stub />
  <div
    class="block-container"
  >
    <h2>
      Block
      <span>
        #42
      </span>
    </h2>
    <div
      class="table"
    >
      <tm-list-item-stub
        class="height"
        dt="Block height"
      >
        <template>
          <span>
            42
          </span>
          <router-link-stub
            to="[object Object]"
          >
            <i
              class="material-icons"
            >
              chevron_left
            </i>
          </router-link-stub>
          <router-link-stub
            to="[object Object]"
          >
            <i
              class="material-icons"
            >
              chevron_right
            </i>
          </router-link-stub>
        </template>
      </tm-list-item-stub>
      <tm-list-item-stub
        dd="2019.04.24 06:03:24  (UTC)"
        dt="Timestamp"
      />
      <tm-list-item-stub
        dt="Transactions"
      >
        <template>
          <a
            class="transactions none"
          >
            <span>
              0 Transactions
            </span>
          </a>
        </template>
      </tm-list-item-stub>
      <tm-list-item-stub
        dt="Proposer"
      >
        <template>
          <span>
            AE699F966DBC31F1EDCDC8A7388A0B9FF827C529
          </span>
        </template>
      </tm-list-item-stub>
      <tm-list-item-stub
        dd="D77DC9D47985955B3E6BAF381C81C4F25A2C00AA75966611B910D02407D62B00"
        dt="Block hash"
      />
      <tm-list-item-stub
        dt="Parent hash"
      >
        <template>
          <router-link-stub
            to="[object Object]"
          >
            13381A34751403CBCB192C34E9C14F30E208A5C08A2A49ADE06766299C30DDDC
          </router-link-stub>
        </template>
      </tm-list-item-stub>
    </div>
  </div>
</div>
`;

exports[`PageBlock Snapshot Snapshot: load Error 1`] = `
<div>
  <app-header-stub />
  <app-not-found-stub />
</div>
`;
